### PR TITLE
Refactor iwyu tool and add a unit test

### PR DESCRIPTION
**Pull request by pfaffe**
Here's a possible way to unittest iwyu_tool. Basicly, I mocked out actually running include-what-you-use.

### CODE REVIEW

File: iwyu_tool_test.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        self.assertSetEqual(
+            set('BAR%d' % n for n in range(100)),
+            set(self.stdout_stub.getvalue()[:-1].split('\n')))
+
+    def test_order_synchronous(self):
+        invocations = [MockInvocation([], '') for _ in range(100)]
+        for n, invocation in enumerate(invocations):
+            invocation.willReturn('BAR%d' % n)
+            from random import random
+            invocation.willBlock(random() / 100)
```
[the remark this answers is not in the document]
Comment (pfaffe) [reply iwyu_tool_test.py]: The blocking bit is to emulated the side effect of real processes in order to shuffle the output essentially.

File: iwyu_tool.py [hunk 2]
```diff
@@ -197,31 +197,21 @@ def slice_compilation_db(compilation_db, selection):
 
     return new_db
 
+def _run_iwyu(invocation, verbose):
```
[the remark this answers is not in the document]
Comment (pfaffe) [reply iwyu_tool.py]: Fortunately, #575 makes this obsolete :)

File: iwyu_tool_test.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/env python
+
+##===--- iwyu_tool_test.py - test for fix_includes.py ---------------------===##
```
Comment (kimgr) [inline iwyu_tool_test.py]: test for iwyu_tool.py

File: iwyu_tool_test.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+                'command': '/usr/bin/clang++ -Iinclude file.cc',
+                'file': 'file.cc'
+            }, self.iwyu_args)
+        self.assertEqual(
+            invocation.command,
+            [iwyu_tool.IWYU_EXECUTABLE, '-foo', '-Iinclude', 'file.cc'])
+        self.assertEqual(invocation.cwd, '/home/user/llvm/build')
+
+    def test_invocation(self):
+        invocation = MockInvocation([], '')
```
Comment (kimgr) [inline iwyu_tool_test.py]: Skip the args now that they're defaulted :)

File: iwyu_tool_test.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        invocation.will_return('BAR')
+        self.assertEqual(self._process_invocations([invocation]), 0)
+        self.assertEqual(self.stdout_stub.getvalue(), 'BAR\n')
+
+    def test_order_asynchronous(self):
+        invocations = [MockInvocation([], '') for _ in range(100)]
```
Comment (kimgr) [inline iwyu_tool_test.py]: No need for args to MockInvocation

File: iwyu_tool_test.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        for n, invocation in enumerate(invocations):
+            invocation.will_return('BAR%d' % n)
+            invocation.will_block(random.random() / 100)
+        self.assertEqual(self._process_invocations(invocations, jobs=100), 0)
+        self.assertSetEqual(
+            set('BAR%d' % n for n in range(100)),
+            set(self.stdout_stub.getvalue()[:-1].splitlines()))
+
+    def test_order_synchronous(self):
+        invocations = [MockInvocation([], '') for _ in range(100)]
```
Comment (kimgr) [inline iwyu_tool_test.py]: No need for args to MockInvocation

File: iwyu_tool_test.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        for n, invocation in enumerate(invocations):
+            invocation.will_return('BAR%d' % n)
+            invocation.will_block(random.random() / 100)
+        self.assertEqual(self._process_invocations(invocations, jobs=1), 0)
+        self.assertEqual(['BAR%d' % n for n in range(100)],
+                         self.stdout_stub.getvalue()[:-1].splitlines())
```
Comment (kimgr) [inline iwyu_tool_test.py]: The `[:-1]` should no longer be necessary

File: iwyu_tool_test.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+    def test_order_asynchronous(self):
+        invocations = [MockInvocation([], '') for _ in range(100)]
+        for n, invocation in enumerate(invocations):
+            invocation.will_return('BAR%d' % n)
+            invocation.will_block(random.random() / 100)
+        self.assertEqual(self._process_invocations(invocations, jobs=100), 0)
+        self.assertSetEqual(
+            set('BAR%d' % n for n in range(100)),
+            set(self.stdout_stub.getvalue()[:-1].splitlines()))
```
Comment (kimgr) [inline iwyu_tool_test.py]: The `[:-1]` should no longer be necessary

File: iwyu_tool_test.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+    def run_iwyu(self, verbose):
+        if self._will_block > 0:
+            time.sleep(self._will_block)
+        return self._will_return
+
+
+class IWYUToolTestBase(unittest.TestCase):
+    def setUp(self):
+        self.iwyu_args = []
```
Comment (kimgr) [inline iwyu_tool_test.py]: You don't really need this as a member yet

File: iwyu_tool.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kimgr) [reply iwyu_tool.py]: Aargh. I can't wait to get rid of multiprocessing...

**Comment by kimgr** [comment]
Thank you!

I fixed some minor style nits(*) and merged this as 2e1e0e16dbcbce4ec9aa21d996db3cc9ed79d1ab.

(*):
- Commit message
- Two blank lines between top-level functions
- Doc comments
- Typo in file header comment iwyu_too.py -> iwyu_tool.py

Thanks for driving this! I'll make an effort to add more relevant tests for compilation database parsing on top of this, because it's surprisingly error prone...

**Pull request closed without merging**